core/src/main/scala/kafka/controller/KafkaController.scala

```
@@ -2225,194 +2223,228 @@ class KafkaController(val config: KafkaConfig,
+        } else {
+          new AlterPartitionResponseData.TopicData()
+            .setTopicName(topicName)
+        }
+        alterPartitionResponse.topics.add(topicResponse)
+
+        partitionResponses.forKeyValue { (tp, errorOrIsr) =>
+          // Add each partition part to the response (new ISR or error)
+          errorOrIsr match {
+            case Left(error) =>
+              topicResponse.partitions.add(
+                new AlterPartitionResponseData.PartitionData()
+                  .setPartitionIndex(tp.partition)
+                  .setErrorCode(error.code))
+            case Right(leaderAndIsr) =>
+              /* Setting the LeaderRecoveryState field is always safe because it will always be the same
+               * as the value set in the request. For version 0, that is always the default RECOVERED
+               * which is ignored when serializing to version 0. For any other version, the
+               * LeaderRecoveryState field is supported.
+               */
+              topicResponse.partitions.add(
+                new AlterPartitionResponseData.PartitionData()
+                  .setPartitionIndex(tp.partition)
+                  .setLeaderId(leaderAndIsr.leader)
+                  .setLeaderEpoch(leaderAndIsr.leaderEpoch)
+                  .setIsr(leaderAndIsr.isr.map(Integer.valueOf).asJava)
+                  .setLeaderRecoveryState(leaderAndIsr.leaderRecoveryState.value)
+                  .setPartitionEpoch(leaderAndIsr.partitionEpoch)
+              )
+          }
+        }
+      }
 
-      Left(partitionResponses)
+      callback(alterPartitionResponse)
     } catch {
       case e: Throwable =>
         error(s"Error when processing AlterPartition for partitions: ${partitionsToAlter.keys.toSeq}", e)
-        Right(Errors.UNKNOWN_SERVER_ERROR)
+        callback(new AlterPartitionResponseData().setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code))
```

Could we let this throw to `processAlterPartition`?

---

Yeah.